### PR TITLE
Add `jsonFallback` option

### PR DESCRIPTION
**Pull request by wooorm**
Actual use case:

- Acorn uses instances of classes for nodes
- This allows them to be serialized, instead of an error about `[object Object]`

Note:

- Naming is up for debate
- There is no tooling in this repo to format the project, but there are configuration files, meaning that prettier in my editor reformats everything I touch, and I can’t match your styleguide

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -44,13 +44,18 @@ export type Value =
...
 
+export interface Options {
+  jsonFallback?: boolean;
+}
+
 /**
  * Convert a value to an ESTree node
  *
  * @param value - The value to convert
+ * @param options
```
Comment (remcohaszing) [inline src/index.ts]:
```suggestion
 * @param options - Additional options to configure the output.
```

File: src/index.ts [hunk 2]
```diff
@@ -44,13 +44,18 @@ export type Value =
...
  * @returns The ESTree node.
  */
-export function valueToEstree(value?: Value): Expression {
+export function valueToEstree(value?: Value, options?: Options): Expression {
```
Comment (remcohaszing) [inline src/index.ts]: I prefer to use a default value over checking if it’s defined later.

```suggestion
export function valueToEstree(value?: Value, options: Options = {}): Expression {
```

Since the type can be basically be anything now, I think it makes to sense to accept an `unknown` value if `options` is defined.

```suggestion
export function valueToEstree(value?: Value): Expression;
export function valueToEstree(value: unknown, options: Options): Expression;
export function valueToEstree(value?: Value, options: Options = {}): Expression {
```

Or perhaps simply:

```suggestion
export function valueToEstree(value?: unknown, options: Options = {}): Expression {
```

and then remove the`Value` type.

File: src/index.ts [hunk 3]
```diff
@@ -168,5 +173,9 @@ export function valueToEstree(value?: Value): Expression {
       })),
     };
   }
+  if (options && options.jsonFallback) {
+    return valueToEstree(JSON.parse(JSON.stringify(value)));
+  }
```
Comment (remcohaszing) [inline src/index.ts]: I’m thinking that instead of using JSON serialization, an option `ignorePrototype` can be used. This would then skip the `isPlainObject` check. In other words:

```diff
- if (isPlainObject(value)) {
+ if (options.ignorePrototype || isPlainObject(value)) {
```

This still allows to process values not serializable using JSON on non-plain objects.

**Review by remcohaszing — CHANGES_REQUESTED** [review]
> Actual use case:
> - Acorn uses instances of classes for nodes

Good to have a practical use case. :)

> - This allows them to be serialized, instead of an error about [object Object]

Hmm, that error could be improved then, but that’s not related to these changes.

> - There is no tooling in this repo to format the project, but there are configuration files, meaning that prettier in my editor reformats everything I touch, and I can’t match your styleguide

I don’t understand. Both `prettier` and `eslint` including dependencies are installed, and the configuration files are there. Your editor should format everything according to the style guide.

**Comment by wooorm** [comment]
> Hmm, that error could be improved then, but that’s not related to these changes.

Yep, my first idea was to have errors based on their path, e.g., `some.field[1].foo`. Would also be good to have

> I don’t understand. Both prettier and eslint including dependencies are installed, and the configuration files are there. Your editor should format everything according to the style guide.

I don‘t have an eslint integration in my editor

File: src/index.ts (hunk 2 above)
Comment (wooorm) [reply src/index.ts]: Alternatively an overload? `unknown` if `OptionsWithJson` or so, and `Value` otherwise?
I’m up for both

File: src/index.ts (hunk 3 above)
Comment (wooorm) [reply src/index.ts]: People can add `toJSON` on their custom objects: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description.
My current method allows that, your method doesn’t.
I do agree that your method is nicer because it “interleaves”, whereas mine is “all or nothing”.
Another alternative is to support that method, perhaps optionally, too?

File: src/index.ts (hunk 2 above)
Comment (remcohaszing) [reply src/index.ts]: Let’s just use `unknown` without the overload. Basically anything is supported except functions anyway.

File: src/index.ts (hunk 3 above)
Comment (remcohaszing) [reply src/index.ts]: I don’t think `toJSON` should be supported, since this library doesn’t serialize to JSON. It could be added later if a practical use case occurs, but for now, I think the `ignorePrototype` option solves the issue in the cleanest way. (Although I’m open to a better name.)

If `toJSON` should be supported at some point, I think it would be better to call the method instead of actually serializing and parsing it from JSON, as this would allow users to specify a `toJSON` method which returns other types that this function can serialize which aren’t supported by JSON.

Comment (wooorm) [reply src/index.ts]: With `toJSON` I mean your last paragraph!

**Comment by remcohaszing** [comment]
> > Hmm, that error could be improved then, but that’s not related to these changes.
> 
> Yep, my first idea was to have errors based on their path, e.g., `some.field[1].foo`. Would also be good to have

That sounds awesome! Maybe I could add the constructor name as well.

> > I don’t understand. Both prettier and eslint including dependencies are installed, and the configuration files are there. Your editor should format everything according to the style guide.
> 
> I don‘t have an eslint integration in my editor

Since it’s prominently used within the JavaScript community, I really recommend you do. :sweat_smile: 

Anyway, aside from that, formatting using Prettier should also just work, and you can run `yarn eslint --fix .` to fix any ESLint issues.

**Comment by wooorm** [comment]
yarn doesn’t work that well 🤷‍♂️

```
$ yarn eslint --fix .
yarn run v1.22.11
$ /Users/tilde/Projects/oss/estree-util-value-to-estree/node_modules/.bin/eslint --fix .
Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db

Why you should do it regularly:
https://github.com/browserslist/browserslist#browsers-data-updating

/Users/tilde/Projects/oss/estree-util-value-to-estree/README.md
  207:7  error  Parsing error: Unexpected token :

/Users/tilde/Projects/oss/estree-util-value-to-estree/src/index.test.ts
  1:1  error  File has too many classes (2). Maximum allowed is 1  max-classes-per-file

/Users/tilde/Projects/oss/estree-util-value-to-estree/src/index.ts
   55:0  error  Missing JSDoc @param "options" description                                                  jsdoc/require-param-description
  176:7  error  Prefer using an optional chain expression instead, as it's more concise and easier to read  @typescript-eslint/prefer-optional-chain
```

---

and editor integrations can often be configured differently that local config files — which is what’s happening for me. I personally don’t think users should have to install other things, unless they’re explicitly documented, and that linting and formatting should also be a script

File: src/index.ts (hunk 3 above)
Comment (remcohaszing) [reply src/index.ts]: I’m not sure what you mean exactly :sweat_smile: 

Can we just implement `ignorePrototype` for now and ignore the `toJSON` part? We can revisit it later if this turns out to be needed.

Comment (wooorm) [reply src/index.ts]: > I’m not sure what you mean exactly 😅

I mean exactly what you wrote:

> If toJSON should be supported at some point, I think it would be better to call the method instead of actually serializing and parsing it from JSON, as this would allow users to specify a toJSON method which returns other types that this function can serialize which aren’t supported by JSON.

**Comment by wooorm** [comment]
Implemented `ignorePrototype`.

Gave it an alternative name: `instanceAsObject`, to signal that “complex objects are turned into simple objects”. Open to alternatives

File: src/index.ts [hunk 4]
```diff
@@ -44,13 +44,18 @@ export type Value =
...
+
 /**
  * Convert a value to an ESTree node
  *
  * @param value - The value to convert
+ * @param options - Additional options to configure the output.
  * @returns The ESTree node.
  */
-export function valueToEstree(value?: Value): Expression {
+export function valueToEstree(value?: unknown, options: Options = {}): Expression {
```
Comment (remcohaszing) [inline src/index.ts:26]: `options` should be passed into all recursive calls to `valueToEstree`. An instance may contain nested properties which may also be class instances which need to be ignored.

File: src/index.ts [hunk 5]
```diff
@@ -44,13 +44,18 @@ export type Value =
 type ValueSet = Set<Value>;
 type ValueMap = Map<Value, Value>;
 
+export interface Options {
+  instanceAsObject?: boolean;
```
Comment (remcohaszing) [inline src/index.ts:16]:
```suggestion
  /**
   * If true, treat objects that have a prototype as plain objects.
   */
  instanceAsObject?: boolean;
```

File: README.md [hunk 6]
```diff
@@ -199,6 +202,43 @@ deepEqual(result, {
     },
   ],
 });
+
+class Point {
+  line: number;
+  column: number;
+  constructor(line: number, column: number) {
```
Comment (remcohaszing) [inline README.md:208]: This is TypeScript syntax inside a JavaScript code block. This causes an ESLint parsing error. Could you either strip the type definitions or change the code block type to `ts`?

File: src/index.ts [hunk 7]
```diff
@@ -44,13 +44,18 @@ export type Value =
...
+}
+
 /**
  * Convert a value to an ESTree node
  *
  * @param value - The value to convert
+ * @param options - Additional options to configure the output.
  * @returns The ESTree node.
  */
-export function valueToEstree(value?: Value): Expression {
```
Comment (remcohaszing) [inline src/index.ts:53]: This makes the Value type unused (except for the export). could you remove it?

**Review by remcohaszing — CHANGES_REQUESTED** [review]
I updated dependencies. The most significant change is that I disabled the ESLint rule `max-classes-per-file`, so you’ll want to merge the main branch first.

**Comment by remcohaszing** [comment]
> yarn doesn’t work that well :man_shrugging:

Looks exactly like the right output :sweat_smile: 

> and editor integrations can often be configured differently that local config files — which is what’s happening for me.

I think it’s best to avoid using formatting integrations or custom configuration for those tools that aren’t part of a project. Some projects don’t even contain any linting or formatting. For these projects you’d want to avoid accidentally formatting every file touched. For those that do, your editor should use the project’s configuration files.

> I personally don’t think users should have to install other things, unless they’re explicitly documented,

I agree, and they don’t. All configurations are there and all tools are installed automatically. One can leverage the CLI tools or their own preferred integrations.

To validate:

```sh
yarn eslint .
yarn prettier --check .
```

To autofix:

```sh
yarn prettier --write .
yarn eslint --fix .
```

> and that linting and formatting should also be a script

This is where we simply disagree. I believe any contributor should be able to use any tool however they like. Many tools contain various command line options that allow users to tweak how they want to run commands, i.e. watch more, test filtering, custom output, or using their own preferred wrapper tool of choice, i.e. editor integrations. IMO scripts tend to get in the way by assuming a tool is run in a certain way, i.e. using specific command line flags, which should be part of a configuration file.

---

TL;DR let’s just agree to disagree. :)

**Comment by wooorm** [comment]
Getting some weird errors on npm install (which needs legacy-peer-deps), and with that:

```
Oops! Something went wrong! :(

ESLint: 7.21.0

TypeError: Error while loading rule 'prettier/prettier': context.getPhysicalFilename is not a function
Occurred while linting /Users/tilde/Projects/oss/estree-util-value-to-estree/jest.config.js
```

**Comment by remcohaszing** [comment]
Thanks! Released as version 1.3.0 :tada:
